src/lib/DateItem.js

```diff
@@ -0,0 +1,121 @@
+// @flow
+
+import Moment from "moment";
+import {Cycles, LiturgicalSeasons, Types} from "../constants";
+import * as Dates from "./Dates";
+
+export opaque type ISOString: string = string;
+
+/**
+ * DateItem Class
+ */
+export default class DateItem {
+  key: string;
+  date: ISOString;
+  type: string;
+  name: string;
+  data: Object;
+  moment: Moment;
+  base: DateItem;
+  _id: number;
+  _stack: number;
+  static latestId: number;
+
+  /**
+   * Create a new DateItem
+   */
+  constructor(item: Object, baseItem: DateItem) {
+    this.key = item.key;
+    this.date = item.moment.toISOString();
+    this.type = item.type;
+    this.name = item.name;
+    this.data = item.data || {};
+    this.data.meta = this.data.meta || {};
+    this.moment = item.moment;
+    this._id = DateItem._incrementId();
+    this._stack = item._stack;
+
+    // The original default item is added to the current item as a `base` property
+    if (baseItem && this._id !== baseItem._id) {
+      this.base = baseItem;
+      this.data = {...{
+          season: baseItem.data.season,
+          calendar: baseItem.data.calendar
+        }, ...this.data};
+      this.data.meta.psalterWeek = this.data.meta.psalterWeek || baseItem.data.meta.psalterWeek;
+    }
+
+    this._adjustTypeInSeasons();
+    this._addLiturgicalCycleMetadata();
+  }
+
+  /**
+   * Special type management in particular seasons
+   */
+  _adjustTypeInSeasons() {
+    // Special type management in the season of LENT
+    if (this.base && this.base.data.season.key === LiturgicalSeasons.LENT) {
+
+      // MEMORIAL or OPT_MEMORIAL that fall on a FERIA
+      // in the SEASON OF LENT are reduced to a COMMEMORATION
+      if (
+        (this.type === Types.MEMORIAL || this.type === Types.OPT_MEMORIAL) &&
+        this.base.type === Types.FERIA
+      ) {
+        this.type = Types.COMMEMORATION
+      }
+
+      // A FEAST occurring in the season of LENT is reduced
+      // to a COMMEMORATION
+      if (
+        this.type === Types.FEAST
+      ) {
+        this.type = Types.COMMEMORATION
+      }
+    }
+  }
+
+  /**
+   * Include liturgical cycle metadata corresponding to the liturgical year
+   */
+  _addLiturgicalCycleMetadata() {
+    const year = this._getLiturgicalStartYear();
+
+    // Formula to calculate lectionary cycle (Year A, B, C)
```

Comment:
`lectionary cycle` → `Sunday cycle`

I believe we `Sunday cycle` is the proper term for this.